pkg/controller/job/job_controller.go

```diff
@@ -808,52 +809,63 @@ func (jm *Controller) syncJob(ctx context.Context, key string) (rErr error) {
+	}
 
-	if feature.DefaultFeatureGate.Enabled(features.JobPodFailurePolicy) {
+	// Given that the Job already has the SuccessCriteriaMet condition, the termination condition already had confirmed in another cycle.
+	// So, the job-controller evaluates the podFailurePolicy only when the Job doesn't have the SuccessCriteriaMet condition.
+	if feature.DefaultFeatureGate.Enabled(features.JobPodFailurePolicy) && jobCtx.successCondition == nil {
 		if failureTargetCondition := findConditionByType(job.Status.Conditions, batch.JobFailureTarget); failureTargetCondition != nil {
-			jobCtx.finishedCondition = newFailedConditionForFailureTarget(failureTargetCondition, jm.clock.Now())
+			jobCtx.failureCondition = newFailedConditionForFailureTarget(failureTargetCondition, jm.clock.Now())
 		} else if failJobMessage := getFailJobMessage(&job, pods); failJobMessage != nil {
 			// Prepare the interim FailureTarget condition to record the failure message before the finalizers (allowing removal of the pods) are removed.
-			jobCtx.finishedCondition = newCondition(batch.JobFailureTarget, v1.ConditionTrue, batch.JobReasonPodFailurePolicy, *failJobMessage, jm.clock.Now())
+			jobCtx.failureCondition = newCondition(batch.JobFailureTarget, v1.ConditionTrue, batch.JobReasonPodFailurePolicy, *failJobMessage, jm.clock.Now())
 		}
 	}
-	if jobCtx.finishedCondition == nil {
+	if jobCtx.failureCondition == nil && jobCtx.successCondition == nil {
 		if exceedsBackoffLimit || pastBackoffLimitOnFailure(&job, pods) {
 			// check if the number of pod restart exceeds backoff (for restart OnFailure only)
 			// OR if the number of failed jobs increased since the last syncJob
-			jobCtx.finishedCondition = newCondition(batch.JobFailed, v1.ConditionTrue, batch.JobReasonBackoffLimitExceeded, "Job has reached the specified backoff limit", jm.clock.Now())
+			jobCtx.failureCondition = newCondition(batch.JobFailed, v1.ConditionTrue, batch.JobReasonBackoffLimitExceeded, "Job has reached the specified backoff limit", jm.clock.Now())
 		} else if jm.pastActiveDeadline(&job) {
-			jobCtx.finishedCondition = newCondition(batch.JobFailed, v1.ConditionTrue, batch.JobReasonDeadlineExceeded, "Job was active longer than specified deadline", jm.clock.Now())
+			jobCtx.failureCondition = newCondition(batch.JobFailed, v1.ConditionTrue, batch.JobReasonDeadlineExceeded, "Job was active longer than specified deadline", jm.clock.Now())
 		} else if job.Spec.ActiveDeadlineSeconds != nil && !jobSuspended(&job) {
 			syncDuration := time.Duration(*job.Spec.ActiveDeadlineSeconds)*time.Second - jm.clock.Since(job.Status.StartTime.Time)
 			logger.V(2).Info("Job has activeDeadlineSeconds configuration. Will sync this job again", "key", key, "nextSyncIn", syncDuration)
 			jm.queue.AddAfter(key, syncDuration)
 		}
 	}
 
-	if isIndexedJob(&job) {
+	if isIndexedJob(&job) && jobCtx.successCondition == nil {
```

Comment:
Thank you for the experiment of splitting the `finishedCondition`. However, it seems this is the only place where we need to discriminate between them, so maybe keeping `finishedCondition` is better, because less code changes, and guarantees that we don't have both !=nil. 

I initially didn't like the long checks where we needed to discriminate between them, but maybe `isSuccessCondition(jobCtx.finishedCondition)` would help keeping the checks short.

Reply:
>  However, it seems this is the only place where we need to discriminate between them,

@mimowo I think that we need to have the previous discriminable in some additional places like Line 816, 861, 1072, 1076, and so on. So, I guess that keeping this `successCondition` would be better. WDYT?

Reply:
Let's see:
- line 816 you can guard using `finishedCondition == nil`
- line 861 you can change `if jobCtx.failureCondition != nil || jobCtx.successCondition != nil {` with `finishedCondition != nil`
- lines 1072 and 1074 are valid, we would need to use `isFailureCondition()` & `isSuccessCondition()`

All in all, there are pros & cons, so I don't have a strong view either way, just pointing out it is something I'm still hestitant about.

Actually, I also don't like that much this block in 1153:
```
	finishedCondition := jobCtx.failureCondition
	if finishedCondition == nil && jobCtx.successCondition != nil {
		finishedCondition = jobCtx.successCondition
	}
	jobFinished := !reachedMaxUncountedPods && jm.enactJobFinished(jobCtx.job, finishedCondition)
```
I think if we go with the split approach (i"m fine) we should pass `jobCtx` into `enactJobFinished`, and `recordJobFinished`.

Reply:
@mimowo That makes sense. I'm also wondering if we should introduce successCondition.
How about asking Aldo's opinion after we improve Line 1153 (passing `jobCtx` to `enactJobFinished` and `recordJobFinished`)?

Reply:
I'm ok with trying to cleanup 1153 first and ask Aldo. If 1153 is cleaned up, I'm ok with both approaches

Reply:
Here is still working.

Reply:
Modified the `enactJobFinished` and `recordJobFinished`.

Reply:
@alculquicondor Please let us know what you think: https://github.com/kubernetes/kubernetes/pull/123412#discussion_r1508757772

Reply:
We decided to revert to the approach not using 2 terminal conditions here: https://github.com/kubernetes/kubernetes/pull/123412#discussion_r1511010610